File: sura_localization/ned_to_enu_imu.py
#!/usr/bin/env python3

from __future__ import annotations

import math

from geometry_msgs.msg import Quaternion, Vector3
from sensor_msgs.msg import Imu
import rclpy
from rclpy.node import Node


NED_TO_ENU_QUATERNION = Quaternion(
    x=math.sqrt(0.5),
    y=math.sqrt(0.5),
    z=0.0,
    w=0.0,
)


class NedToEnuImu(Node):
    def __init__(self) -> None:
        super().__init__("ned_to_enu_imu")

        self.declare_parameter("input_topic", "/cirtesub/sensors/imu")
        self.declare_parameter("output_topic", "/cirtesub/sensors/imu_enu")
        self.declare_parameter("frame_id", "cirtesub/IMU")

        input_topic = str(self.get_parameter("input_topic").value)
        output_topic = str(self.get_parameter("output_topic").value)

        self._publisher = self.create_publisher(Imu, output_topic, 10)
        self._subscription = self.create_subscription(
            Imu,
            input_topic,
            self._imu_callback,
            10,
        )

        self.get_logger().info(f"Converting NED IMU {input_topic} to ENU {output_topic}")

    def _imu_callback(self, msg: Imu) -> None:
        converted = Imu()
        converted.header = msg.header
        converted.header.frame_id = str(self.get_parameter("frame_id").value)

        converted.orientation = self._rotate_orientation(msg.orientation)
        converted.orientation_covariance = self._transform_3x3_covariance(
            msg.orientation_covariance
        )
        converted.angular_velocity = self._transform_vector(msg.angular_velocity)
        converted.angular_velocity_covariance = self._transform_3x3_covariance(
            msg.angular_velocity_covariance
        )
        converted.linear_acceleration = self._transform_vector(msg.linear_acceleration)
        converted.linear_acceleration_covariance = self._transform_3x3_covariance(
            msg.linear_acceleration_covariance
        )

        self._publisher.publish(converted)

    @staticmethod
    def _transform_vector(vector: Vector3) -> Vector3:
        converted = Vector3()
        converted.x = vector.y
        converted.y = vector.x
        converted.z = -vector.z
        return converted

    @staticmethod
    def _rotate_orientation(orientation: Quaternion) -> Quaternion:
        return NedToEnuImu._normalize_quaternion(
            NedToEnuImu._multiply_quaternions(NED_TO_ENU_QUATERNION, orientation)
        )

    @staticmethod
    def _multiply_quaternions(left: Quaternion, right: Quaternion) -> Quaternion:
        result = Quaternion()
        result.w = (
            left.w * right.w
            - left.x * right.x
            - left.y * right.y
            - left.z * right.z
        )
        result.x = (
            left.w * right.x
            + left.x * right.w
            + left.y * right.z
            - left.z * right.y
        )
        result.y = (
            left.w * right.y
            - left.x * right.z
            + left.y * right.w
            + left.z * right.x
        )
        result.z = (
            left.w * right.z
            + left.x * right.y
            - left.y * right.x
            + left.z * right.w
        )
        return result

    @staticmethod
    def _normalize_quaternion(quaternion: Quaternion) -> Quaternion:
        norm = math.sqrt(
            quaternion.x * quaternion.x
            + quaternion.y * quaternion.y
            + quaternion.z * quaternion.z
            + quaternion.w * quaternion.w
        )
        if norm <= 0.0:
            return Quaternion(w=1.0)

        quaternion.x /= norm
        quaternion.y /= norm
        quaternion.z /= norm
        quaternion.w /= norm
        return quaternion

    @staticmethod
    def _transform_3x3_covariance(covariance: list[float]) -> list[float]:
        transform = (
            (0.0, 1.0, 0.0),
            (1.0, 0.0, 0.0),
            (0.0, 0.0, -1.0),
        )
        return [
            sum(
                transform[row][i] * covariance[i * 3 + j] * transform[col][j]
                for i in range(3)
                for j in range(3)
            )
            for row in range(3)
            for col in range(3)
        ]


def main(args=None) -> None:
    rclpy.init(args=args)
    node = NedToEnuImu()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/ned_to_enu_imu.cpp
#include <geometry_msgs/msg/quaternion.hpp>
#include <geometry_msgs/msg/vector3.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>

#include <array>
#include <cmath>
#include <memory>
#include <string>

namespace
{

geometry_msgs::msg::Vector3 transform_vector(const geometry_msgs::msg::Vector3 & vector)
{
  geometry_msgs::msg::Vector3 converted;
  converted.x = vector.x;
  converted.y = -vector.y;
  converted.z = -vector.z;
  return converted;
}

geometry_msgs::msg::Quaternion multiply_quaternions(
  const geometry_msgs::msg::Quaternion & left,
  const geometry_msgs::msg::Quaternion & right)
{
  geometry_msgs::msg::Quaternion result;
  result.w = left.w * right.w - left.x * right.x - left.y * right.y - left.z * right.z;
  result.x = left.w * right.x + left.x * right.w + left.y * right.z - left.z * right.y;
  result.y = left.w * right.y - left.x * right.z + left.y * right.w + left.z * right.x;
  result.z = left.w * right.z + left.x * right.y - left.y * right.x + left.z * right.w;
  return result;
}

geometry_msgs::msg::Quaternion normalize_quaternion(geometry_msgs::msg::Quaternion quaternion)
{
  const double norm = std::sqrt(
    quaternion.x * quaternion.x +
    quaternion.y * quaternion.y +
    quaternion.z * quaternion.z +
    quaternion.w * quaternion.w);
  if (norm <= 0.0) {
    geometry_msgs::msg::Quaternion identity;
    identity.w = 1.0;
    return identity;
  }
  quaternion.x /= norm;
  quaternion.y /= norm;
  quaternion.z /= norm;
  quaternion.w /= norm;
  return quaternion;
}

geometry_msgs::msg::Quaternion rotate_orientation(
  const geometry_msgs::msg::Quaternion & orientation)
{
  geometry_msgs::msg::Quaternion ned_to_enu;
  ned_to_enu.x = std::sqrt(0.5);
  ned_to_enu.y = std::sqrt(0.5);
  ned_to_enu.z = 0.0;
  ned_to_enu.w = 0.0;

  geometry_msgs::msg::Quaternion frd_to_flu;
  frd_to_flu.x = 1.0;
  frd_to_flu.y = 0.0;
  frd_to_flu.z = 0.0;
  frd_to_flu.w = 0.0;

  return normalize_quaternion(
    multiply_quaternions(multiply_quaternions(ned_to_enu, orientation), frd_to_flu));
}

std::array<double, 9> transform_covariance(const std::array<double, 9> & covariance)
{
  constexpr double transform[3][3] = {
    {1.0, 0.0, 0.0},
    {0.0, -1.0, 0.0},
    {0.0, 0.0, -1.0},
  };

  std::array<double, 9> converted{};
  for (int row = 0; row < 3; ++row) {
    for (int col = 0; col < 3; ++col) {
      double value = 0.0;
      for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
          value += transform[row][i] * covariance[i * 3 + j] * transform[col][j];
        }
      }
      converted[row * 3 + col] = value;
    }
  }
  return converted;
}

class NedToEnuImu : public rclcpp::Node
{
public:
  NedToEnuImu()
  : Node("ned_to_enu_imu")
  {
    declare_parameter<std::string>("input_topic", "/cirtesub/sensors/imu");
    declare_parameter<std::string>("output_topic", "/cirtesub/sensors/imu_enu");
    declare_parameter<std::string>("frame_id", "cirtesub/IMU");
    declare_parameter<double>("orientation_yaw_stddev_deg", -1.0);

    const auto input_topic = get_parameter("input_topic").as_string();
    const auto output_topic = get_parameter("output_topic").as_string();
    const double yaw_stddev_deg = get_parameter("orientation_yaw_stddev_deg").as_double();
    if (yaw_stddev_deg >= 0.0) {
      constexpr double pi = 3.14159265358979323846;
      const double yaw_stddev_rad = yaw_stddev_deg * pi / 180.0;
      orientation_yaw_variance_ = yaw_stddev_rad * yaw_stddev_rad;
    }

    publisher_ = create_publisher<sensor_msgs::msg::Imu>(output_topic, 10);
    subscription_ = create_subscription<sensor_msgs::msg::Imu>(
      input_topic,
      10,
      [this](const sensor_msgs::msg::Imu::SharedPtr msg) {
        on_imu(*msg);
      });

    RCLCPP_INFO(
      get_logger(),
      "Converting NED IMU %s to ENU %s",
      input_topic.c_str(),
      output_topic.c_str());
  }

private:
  void on_imu(const sensor_msgs::msg::Imu & msg)
  {
    sensor_msgs::msg::Imu converted;
    converted.header = msg.header;
    converted.header.frame_id = get_parameter("frame_id").as_string();
    converted.orientation = rotate_orientation(msg.orientation);
    converted.orientation_covariance = transform_covariance(msg.orientation_covariance);
    if (orientation_yaw_variance_ >= 0.0) {
      converted.orientation_covariance[8] = orientation_yaw_variance_;
    }
    converted.angular_velocity = transform_vector(msg.angular_velocity);
    converted.angular_velocity_covariance = transform_covariance(msg.angular_velocity_covariance);
    converted.linear_acceleration = transform_vector(msg.linear_acceleration);
    converted.linear_acceleration_covariance =
      transform_covariance(msg.linear_acceleration_covariance);
    publisher_->publish(converted);
  }

  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr publisher_;
  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr subscription_;
  double orientation_yaw_variance_{-1.0};
};

}  // namespace

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<NedToEnuImu>());
  rclcpp::shutdown();
  return 0;
}
